### PR TITLE
Secrets 🔐 PR 3 - Integrate secret handling with db-details

### PR DESCRIPTION
**Pull request by jeff303**
Add helper function to database model namespace to "swap out" secret values with inserted IDs

Calling this from the pre-update fn

Add pre-insert fn for database model to perform the same

Adding test to database model test ns

Part of epic: #18036

### CODE REVIEW

**Comment by jeff303** [comment]
This is PR 3 for implementing secret support. Builds on top of #17649.

File: src/metabase/models/secret.clj [hunk 1]
```diff
@@ -25,6 +27,46 @@
...
+  bump-version-keys
+  [:kind :source :value])
+
+(defn latest-for-id
+  "Returns the latest Secret instance for the given `id` (meaning the one with the highest `version`)."
+  {:added "0.41.0"}
+  [id]
+  (db/select-one Secret :id id {:order-by [[:version :desc]]}))
+
+(defn upsert-secret-value!
```
Comment (pawit-metabase) [inline src/metabase/models/secret.clj:41]: What's the plan for deletion of secrets/secret versions?

Scenario 1 - I added a data source and decided to remove it from Metabase. I expect the password to be deleted from the DB as well.

Scenario 2 - If I did not enable encryption at first but then later turn it on, I expect the old unencrypted value to be deleted from the DB.

File: src/metabase/models/database.clj [hunk 2]
```diff
@@ -69,10 +71,60 @@
...
+                ;; saved-details         (db/select-field :details Database id)
+                 updated-details      (reduce-kv (partial handle-db-details-secret-prop! database)
+                                                 details
+                                                 conn-secrets-by-name)]
+           (assoc database :details updated-details))
+
+          :else ; no details populated; do nothing
+          database)))
+
 ;; TODO - this logic would make more sense in post-update if such a method existed
```
Comment (pawit-metabase) [inline src/metabase/models/database.clj:124]: this comment should be moved down as the `(handle-secrets-changes database)` part does not belong in the hypothetical `post-update`

File: src/metabase/models/secret.clj (hunk 1 above)
Comment (jeff303) [reply src/metabase/models/secret.clj:41]: Good points.  I will address in a separate PR to keep these fairly bite sized.

File: src/metabase/models/database.clj (hunk 2 above)
Comment (jeff303) [reply src/metabase/models/database.clj]: OK, I think I see what you mean.  See update in latest commit.

Comment (pawit-metabase) [reply src/metabase/models/database.clj]: Yep. That's what I meant.

I'm not yet familiar enough with the code base to review the rest of the PRs so I won't be approving here.

**Review by camsaul — APPROVED** [review]
works for me

**Pull request merged**